cmd/notify-mailer/main.go

```diff
@@ -231,87 +254,138 @@ func emailsForReg(id int, dbMap dbSelector) ([]string, error) {
+	id int64
+
+	// Extra is exported so the contents can be accessed to fill template
+	// variables referenced in the email body.
 	Extra map[string]string
 }
 
-// emailToRecipientMap maps from an email address to a list of recipients with
-// that email address.
-type emailToRecipientMap map[string][]recipient
+// addressesToRecipientsMap maps from an email address to a list of `recipient`s
+// with the same email address.
+type addressesToRecipientsMap map[string][]recipient
 
-// readRecipientsList reads a CSV filename and parses that file into a list of
-// recipient structs. It puts any columns after the first into a per-recipient
-// map from column name -> value.
-func readRecipientsList(filename string) ([]recipient, error) {
-	f, err := os.Open(filename)
+// parseRecipientList parses the contents of a recipient list file into a list
+// of `recipient` objects. All columns after `id` into a per-recipient map from
+// column name -> value.
+func parseRecipientList(recipientList *csv.Reader) ([]recipient, error) {
+	entries, err := recipientList.Read()
 	if err != nil {
 		return nil, err
 	}
-	reader := csv.NewReader(f)
-	record, err := reader.Read()
-	if err != nil {
-		return nil, err
-	}
-	if len(record) == 0 {
-		return nil, fmt.Errorf("no entries in CSV")
-	}
-	if record[0] != "id" {
-		return nil, fmt.Errorf("first field of CSV input must be an ID.")
+
+	if entries[0] != "id" {
+		return nil, fmt.Errorf(
+			"the first column of the header row must be \"id\", got: %q", entries[0])
 	}
-	var columnNames []string
-	for _, v := range record[1:] {
-		columnNames = append(columnNames, strings.TrimSpace(v))
+
+	// Collect column names used for the `Extra` field of the `recipient`
+	// object. All columns after `id` are considered `Extra`.
+	header := entries[1:]
+	var extraColNames []string
+	for _, entry := range header {
+		colName := strings.TrimSpace(entry)
+		if len(colName) == 0 {
+			return nil, fmt.Errorf(
+				"header: %q contains duplicate or trailing delimiter", header)
+		}
+		extraColNames = append(extraColNames, colName)
 	}
 
-	results := []recipient{}
+	var recipients []recipient
 	for {
-		record, err := reader.Read()
-		if err == io.EOF {
-			if len(results) == 0 {
-				return nil, fmt.Errorf("no entries after the header in CSV")
-			}
-			return results, nil
-		}
+		entry, err := recipientList.Read()
 		if err != nil {
+			if errors.Is(err, io.EOF) {
+				// Finished parsing the file.
+				break
+			}
 			return nil, err
 		}
-		if len(record) == 0 {
-			return nil, fmt.Errorf("empty line in CSV")
-		}
-		if len(record) != len(columnNames)+1 {
-			return nil, fmt.Errorf("Number of columns in CSV line didn't match header columns."+
-				" Got %d, expected %d. Line: %v", len(record), len(columnNames)+1, record)
-		}
-		id, err := strconv.Atoi(record[0])
+
+		// Attempt to coerce the value provided for `id` to a valid registration
+		// ID (int64).
+		entryID := entry[0]
+		id, err := strconv.ParseInt(entryID, 10, 64)
```

Comment:
"coerce" usually means type coercion. What we're doing here is parsing. But we don't really need a comment about that - it's inherent in the function call `ParseInt`.